File: MacFileTransfer/MacFileTransfer/Networking/FileTransferEngine.swift
import Foundation
import CryptoKit

protocol FileTransferEngineDelegate: AnyObject {
    func engine(_ engine: FileTransferEngine, didUpdateProgress progress: TransferProgress)
    func engine(_ engine: FileTransferEngine, didCompleteTransfer requestId: String, fileId: String, success: Bool)
    func engine(_ engine: FileTransferEngine, didReceiveChunkAck fileId: String, chunkNumber: Int)
    func engine(_ engine: FileTransferEngine, requiresSend message: Data, toDeviceId: UUID)
}

final class FileTransferEngine {

    static let chunkSize = 2 * 1024 * 1024   // 2 MB
    weak var delegate: FileTransferEngineDelegate?

    private var sendCheckpoints: [String: Int]  = [:]
    private var sendPaused:      [String: Bool] = [:]
    private var sendCancelled:   [String: Bool] = [:]
    private var receiveHandles:  [String: FileHandle] = [:]
    private var receivePaths:    [String: URL]  = [:]
    private var receiveDests:    [String: URL]  = [:]
    private var receiveProgress: [String: TransferProgress] = [:]
    private let queue = DispatchQueue(label: "com.filetransfer.engine", qos: .userInitiated)

    // MARK: - Send

    func sendFiles(files: [URL], requestId: String, toDeviceId: UUID) {
        queue.async {
            for url in files {
                guard !self.sendCancelled[requestId, default: false] else { break }
                self.sendFile(url: url, requestId: requestId, toDeviceId: toDeviceId)
            }
        }
    }

    private func sendFile(url: URL, requestId: String, toDeviceId: UUID) {
        guard let data = try? Data(contentsOf: url, options: .mappedIfSafe) else { return }
        let fileId      = UUID().uuidString
        let fileName    = url.lastPathComponent
        let totalChunks = Int(ceil(Double(data.count) / Double(Self.chunkSize)))
        let startChunk  = sendCheckpoints[fileId, default: 0]
        sendPaused[fileId] = false; sendCancelled[fileId] = false

        var progress = TransferProgress(requestId: requestId, fileId: fileId, fileName: fileName,
                                        totalBytes: Int64(data.count),
                                        transferredBytes: Int64(startChunk * Self.chunkSize))
        var lastUpdate = Date(); var bytesSince: Int64 = 0

        for chunkNum in startChunk..<totalChunks {
            while sendPaused[fileId, default: false] { Thread.sleep(forTimeInterval: 0.2) }
            guard !sendCancelled[fileId, default: false] else { return }

            let start = chunkNum * Self.chunkSize
            let end   = min(start + Self.chunkSize, data.count)
            let chunk = Data(data[start..<end])

            let msg = MessageProtocol.makeFileChunk(fileId: fileId, requestId: requestId,
                                                    chunkNumber: chunkNum, totalChunks: totalChunks,
                                                    data: chunk, fileName: fileName)
            guard let encoded = try? JSONEncoder().encode(msg) else { continue }
            var len = UInt32(encoded.count).bigEndian
            var pkt = Data(bytes: &len, count: 4); pkt.append(encoded)
            delegate?.engine(self, requiresSend: pkt, toDeviceId: toDeviceId)

            waitForAck(fileId: fileId, expectedChunk: chunkNum)

            bytesSince += Int64(chunk.count); progress.transferredBytes += Int64(chunk.count)
            let elapsed = Date().timeIntervalSince(lastUpdate)
            if elapsed >= 0.5 {
                progress.speedBytesPerSecond = Double(bytesSince) / elapsed
                bytesSince = 0; lastUpdate = Date()
            }
            DispatchQueue.main.async { self.delegate?.engine(self, didUpdateProgress: progress) }
        }

        let checksum = SHA256.hash(data: data).map { String(format: "%02x", $0) }.joined()
        let complete = MessageProtocol.makeTransferComplete(fileId: fileId, requestId: requestId,
                                                            checksum: checksum, success: true)
        if let encoded = try? JSONEncoder().encode(complete) {
            var len = UInt32(encoded.count).bigEndian
            var pkt = Data(bytes: &len, count: 4); pkt.append(encoded)
            delegate?.engine(self, requiresSend: pkt, toDeviceId: toDeviceId)
        }
        sendCheckpoints.removeValue(forKey: fileId)
    }

    private func waitForAck(fileId: String, expectedChunk: Int) {
        let timeout = Date().addingTimeInterval(10)
        while sendCheckpoints[fileId, default: -1] < expectedChunk {
            if Date() > timeout || sendCancelled[fileId, default: false] { break }
            Thread.sleep(forTimeInterval: 0.01)
        }
    }

    func handleChunkAck(_ msg: ChunkAckMessage) {
        sendCheckpoints[msg.fileId] = msg.chunkNumber
        delegate?.engine(self, didReceiveChunkAck: msg.fileId, chunkNumber: msg.chunkNumber)
    }

    func pauseSend(fileId: String)  { sendPaused[fileId]    = true  }
    func resumeSend(fileId: String) { sendPaused[fileId]    = false }
    func cancelSend(fileId: String) { sendCancelled[fileId] = true; sendPaused[fileId] = false }

    // MARK: - Receive

    func prepareReceive(fileId: String, fileName: String, totalBytes: Int64, destinationFolder: URL) {
        let tmpURL  = destinationFolder.appendingPathComponent(".\(fileId).tmp")
        let dstURL  = destinationFolder.appendingPathComponent(fileName)
        FileManager.default.createFile(atPath: tmpURL.path, contents: nil)
        guard let handle = try? FileHandle(forWritingTo: tmpURL) else { return }
        receiveHandles[fileId]  = handle
        receivePaths[fileId]    = tmpURL
        receiveDests[fileId]    = dstURL
        receiveProgress[fileId] = TransferProgress(requestId: "", fileId: fileId,
                                                    fileName: fileName, totalBytes: totalBytes)
    }

    func handleIncomingChunk(_ msg: FileChunkMessage, fromDeviceId: UUID) {
        guard let chunkData = Data(base64Encoded: msg.data),
              let handle    = receiveHandles[msg.fileId] else { return }
        handle.seekToEndOfFile(); handle.write(chunkData)

        if var p = receiveProgress[msg.fileId] {
            p.transferredBytes += Int64(chunkData.count)
            let elapsed = Date().timeIntervalSince(p.startedAt)
            p.speedBytesPerSecond = elapsed > 0 ? Double(p.transferredBytes) / elapsed : 0
            receiveProgress[msg.fileId] = p
            DispatchQueue.main.async { self.delegate?.engine(self, didUpdateProgress: p) }
        }

        let ack = MessageProtocol.makeChunkAck(fileId: msg.fileId, chunkNumber: msg.chunkNumber)
        if let encoded = try? JSONEncoder().encode(ack) {
            var len = UInt32(encoded.count).bigEndian
            var pkt = Data(bytes: &len, count: 4); pkt.append(encoded)
            delegate?.engine(self, requiresSend: pkt, toDeviceId: fromDeviceId)
        }
    }

    func handleTransferComplete(_ msg: TransferCompleteMessage) {
        guard let handle = receiveHandles[msg.fileId],
              let tmpURL = receivePaths[msg.fileId],
              let dstURL = receiveDests[msg.fileId] else { return }
        handle.closeFile()

        guard let data = try? Data(contentsOf: tmpURL) else {
            finalize(fileId: msg.fileId, success: false, requestId: msg.requestId); return
        }
        let computed = SHA256.hash(data: data).map { String(format: "%02x", $0) }.joined()
        if computed == msg.checksum { try? FileManager.default.moveItem(at: tmpURL, to: dstURL) }
        else                        { try? FileManager.default.removeItem(at: tmpURL) }
        finalize(fileId: msg.fileId, success: computed == msg.checksum, requestId: msg.requestId)
    }

    private func finalize(fileId: String, success: Bool, requestId: String) {
        receiveHandles.removeValue(forKey: fileId); receivePaths.removeValue(forKey: fileId)
        receiveDests.removeValue(forKey: fileId);   receiveProgress.removeValue(forKey: fileId)
        DispatchQueue.main.async {
            self.delegate?.engine(self, didCompleteTransfer: requestId, fileId: fileId, success: success)
        }
    }
}

File: AndroidFileTransfer/app/src/main/java/com/filetransfer/network/FileTransferEngine.kt
package com.filetransfer.network

import android.util.Base64
import com.filetransfer.model.TransferProgress
import kotlinx.coroutines.*
import java.io.File
import java.io.FileOutputStream
import java.security.MessageDigest
import java.util.UUID

private const val CHUNK_SIZE = 2 * 1024 * 1024   // 2 MB — must match Mac

interface FileTransferEngineListener {
    fun onProgress(progress: TransferProgress)
    fun onComplete(requestId: String, fileId: String, success: Boolean)
    fun onRequiresSend(data: ByteArray, toDeviceId: UUID)
}

class FileTransferEngine {

    var listener: FileTransferEngineListener? = null
    var receiveDirectory: File? = null

    private val scope = CoroutineScope(Dispatchers.IO + SupervisorJob())
    private val sendCheckpoints = mutableMapOf<String, Int>()
    private val sendPaused      = mutableMapOf<String, Boolean>()
    private val sendCancelled   = mutableMapOf<String, Boolean>()
    private val receiveStreams   = mutableMapOf<String, FileOutputStream>()
    private val receivePaths    = mutableMapOf<String, File>()
    private val receiveDests    = mutableMapOf<String, File>()
    private val receiveProgress = mutableMapOf<String, TransferProgress>()

    // MARK: - Send

    fun sendFiles(files: List<File>, requestId: String, toDeviceId: UUID) {
        scope.launch {
            for (file in files) {
                if (sendCancelled[requestId] == true) break
                sendFile(file, requestId, toDeviceId)
            }
        }
    }

    private suspend fun sendFile(file: File, requestId: String, toDeviceId: UUID) {
        val data        = file.readBytes()
        val fileId      = UUID.randomUUID().toString()
        val fileName    = file.name
        val totalChunks = Math.ceil(data.size.toDouble() / CHUNK_SIZE).toInt().coerceAtLeast(1)
        sendPaused[fileId] = false; sendCancelled[fileId] = false

        var progress = TransferProgress(requestId, fileId, fileName, data.size.toLong())
        var lastUpdate = System.currentTimeMillis(); var bytesSince = 0L

        for (chunkNum in 0 until totalChunks) {
            while (sendPaused[fileId] == true) delay(200)
            if (sendCancelled[fileId] == true) return

            val start = chunkNum * CHUNK_SIZE
            val end   = minOf(start + CHUNK_SIZE, data.size)
            val chunk = data.copyOfRange(start, end)
            val encoded = Base64.encodeToString(chunk, Base64.NO_WRAP)

            val msg = MessageProtocol.makeFileChunk(fileId, requestId, chunkNum,
                totalChunks, encoded, fileName, chunk.size)
            val framed = MessageProtocol.frame(MessageProtocol.toJson(msg))
            listener?.onRequiresSend(framed, toDeviceId)

            // Wait for ACK (simple polling with 10s timeout)
            val timeout = System.currentTimeMillis() + 10_000
            while ((sendCheckpoints[fileId] ?: -1) < chunkNum) {
                if (System.currentTimeMillis() > timeout || sendCancelled[fileId] == true) break
                delay(10)
            }

            bytesSince += chunk.size
            progress = progress.copy(transferredBytes = progress.transferredBytes + chunk.size)
            val elapsed = (System.currentTimeMillis() - lastUpdate) / 1000.0
            if (elapsed >= 0.5) {
                progress = progress.copy(speedBytesPerSecond = bytesSince / elapsed)
                bytesSince = 0; lastUpdate = System.currentTimeMillis()
            }
            withContext(Dispatchers.Main) { listener?.onProgress(progress) }
        }

        // Send completion with SHA-256
        val checksum = sha256hex(data)
        val complete = MessageProtocol.makeTransferComplete(fileId, requestId, checksum, true)
        listener?.onRequiresSend(MessageProtocol.frame(MessageProtocol.toJson(complete)), toDeviceId)
        sendCheckpoints.remove(fileId)
    }

    fun handleChunkAck(msg: ChunkAckMessage) { sendCheckpoints[msg.fileId] = msg.chunkNumber }
    fun pauseSend(fileId: String)  { sendPaused[fileId]    = true  }
    fun resumeSend(fileId: String) { sendPaused[fileId]    = false }
    fun cancelSend(fileId: String) { sendCancelled[fileId] = true; sendPaused[fileId] = false }

    // MARK: - Receive

    fun handleIncomingChunk(msg: FileChunkMessage, fromDeviceId: UUID) {
        // Auto-register on first chunk (mirrors Mac behaviour)
        if (!receiveStreams.containsKey(msg.fileId)) {
            val dir = receiveDirectory ?: return
            val tmp  = File(dir, ".${msg.fileId}.tmp")
            val dest = File(dir, msg.fileName)
            receiveStreams[msg.fileId]   = FileOutputStream(tmp)
            receivePaths[msg.fileId]     = tmp
            receiveDests[msg.fileId]     = dest
            val totalBytes = msg.totalChunks.toLong() * msg.chunkSize
            receiveProgress[msg.fileId]  = TransferProgress("", msg.fileId, msg.fileName, totalBytes)
        }

        val chunk = Base64.decode(msg.data, Base64.NO_WRAP)
        receiveStreams[msg.fileId]?.write(chunk)

        receiveProgress[msg.fileId]?.let { p ->
            val updated = p.copy(transferredBytes = p.transferredBytes + chunk.size,
                speedBytesPerSecond = if ((System.currentTimeMillis() - p.startedAt) > 0)
                    p.transferredBytes.toDouble() / ((System.currentTimeMillis() - p.startedAt) / 1000.0) else 0.0)
            receiveProgress[msg.fileId] = updated
            scope.launch(Dispatchers.Main) { listener?.onProgress(updated) }
        }

        // Send ACK
        val ack = MessageProtocol.makeChunkAck(msg.fileId, msg.chunkNumber)
        listener?.onRequiresSend(MessageProtocol.frame(MessageProtocol.toJson(ack)), fromDeviceId)
    }

    fun handleTransferComplete(msg: TransferCompleteMessage) {
        val stream = receiveStreams[msg.fileId] ?: return
        val tmpFile = receivePaths[msg.fileId]  ?: return
        val dstFile = receiveDests[msg.fileId]  ?: return
        stream.close()

        val data = tmpFile.readBytes()
        val computed = sha256hex(data)
        val success = computed == msg.checksum
        if (success) tmpFile.renameTo(dstFile) else tmpFile.delete()

        receiveStreams.remove(msg.fileId); receivePaths.remove(msg.fileId)
        receiveDests.remove(msg.fileId);  receiveProgress.remove(msg.fileId)
        scope.launch(Dispatchers.Main) { listener?.onComplete(msg.requestId, msg.fileId, success) }
    }

    // MARK: - Helpers

    private fun sha256hex(data: ByteArray): String {
        val md = MessageDigest.getInstance("SHA-256")
        return md.digest(data).joinToString("") { "%02x".format(it) }
    }
}
